src/engine.ml

```diff
@@ -825,35 +825,22 @@ let incoming_control_client config rng session channel now op data =
       Log.info (fun m -> m "channel %d is established now!!!" channel.keyid);
       let ip_config = ip_from_config config' in
       (Some (`Established ip_config), config', { channel with channel_st }, [])
-  | Established (tls, keys), Packet.Control -> (
+  | Established (tls, keys), Packet.Control ->
       let open Result.Syntax in
       let+ tls', d = incoming_tls_without_reply tls data in
-      match d with
-      | Some d ->
-          let d = Cstruct.to_string d in
-          if
-            String.equal d "RESTART\000"
-            || String.equal d "EXIT\000" || String.equal d "HALT\000"
-          then
-            let () = Log.warn (fun m -> m "Received control message: %S" d) in
-            ( Some `Exit,
-              config,
-              { channel with channel_st = Established (tls', keys) },
-              [] )
-          else
-            let () =
-              Log.warn (fun m -> m "Received unknown control message: %S" d)
-            in
-            ( None,
-              config,
-              { channel with channel_st = Established (tls', keys) },
-              [] )
-      | None ->
-          (* a bit odd to return [None] ip_config *)
-          ( None,
-            config,
-            { channel with channel_st = Established (tls', keys) },
-            [] ))
+      let channel = { channel with channel_st = Established (tls', keys) } in
+      let act =
+        match d with
+        | None -> None
+        | Some d -> (
+            let d = Cstruct.to_string d in
+            match Cc_message.parse d with
```

Comment:
Here I can think of an Option.iter...

Reply:
There is unfortunately no `Option.iter_none`, and it's in the `None`-case we want to log a warning :/